src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes/ApiControllerCodeFixer.cs

```diff
@@ -0,0 +1,89 @@
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = "UA0013 CodeFix Provider")]
+    public class ApiControllerCodeFixer : CodeFixProvider
+    {
+        public const string GoodNamespace = "Microsoft.AspNetCore.Mvc";
+        public const string GoodClassName = "Controller";
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ApiControllerAnalyzer.DiagnosticId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            if (root is null)
+            {
+                return;
+            }
+
+            var node = root.FindNode(context.Span, false, true);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            // Register a code action that will invoke the fix.
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    CodeFixResources.ApiControllerTitle,
+                    cancellationToken => ReplaceBaseClass(context.Document, node, cancellationToken),
+                    nameof(CodeFixResources.ApiControllerTitle)),
+                context.Diagnostics);
+        }
+
+        private async Task<Document> ReplaceBaseClass(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var project = document.Project;
+            var slnEditor = new SolutionEditor(project.Solution);
+
+            var docEditor = await slnEditor.GetDocumentEditorAsync(document.Id, cancellationToken).ConfigureAwait(false);
+
+            ReplaceController(node, docEditor);
+
+            return docEditor.GetChangedDocument();
+        }
+
+        private static void ReplaceController(SyntaxNode node, DocumentEditor docEditor)
+        {
+            if (IsQualifiedNameSyntax(node))
+            {
+                var namespaceIdentifier = docEditor.Generator.IdentifierName(GoodNamespace);
+                var classNameIdentifier = docEditor.Generator.IdentifierName(GoodClassName);
+                var controllerIdentifierSyntax = docEditor.Generator.QualifiedName(namespaceIdentifier, classNameIdentifier);
+                docEditor.ReplaceNode(node, controllerIdentifierSyntax);
+            }
+            else
+            {
+                var controllerIdentifierSyntax = docEditor.Generator.IdentifierName($"{GoodNamespace}.{GoodClassName}");
+                docEditor.ReplaceNode(node, controllerIdentifierSyntax);
+            }
```

Comment:
It would be nice to add a `using` statement and use a simplifier annotation so that users end up with `: Controller` instead of `: Microsoft.AspNetCore.Mvc.Controller` which is a bit odd.

Reply:
If using the generator, you can have it auto-simplify to whatever is in scope with the given usings

Reply:
I was thinking about that too but decided to go with this simple approach pending our design review.

One idea surfaced is the idea of building a more composable and abstract approach. If we use that approach, then the code here would just be part of the overall solution and could become the BaseTypeMapper or something similar. To achieve the idea you're mentioning I think we would combine the BaseTypeMapper with the NamespaceMapper.